src/dune_rules/coq/coq_lib.ml

```diff
@@ -407,6 +413,150 @@ module DB = struct
+    | Ok Unix.S_DIR -> (
+      let* dir_contents =
+        Fs_memo.dir_contents (Path.as_outside_build_dir_exn dir)
+      in
+      match dir_contents with
+      | Ok x ->
+        let dir_files =
+          List.filter_map (Fs_cache.Dir_contents.to_list x)
+            ~f:(fun (file, kind) ->
+              match kind with
+              | Unix.S_DIR -> Some file
+              | _ -> None)
+        in
+        let prefix_entries = Coq_lib_name.Map.singleton name dir in
+        let+ subdirs_entries =
+          List.map dir_files ~f:(fun file ->
+              let name = Coq_lib_name.append name file in
+              let dir = Path.append_local dir (Path.Local.of_string file) in
+              subdirectory_map name dir)
+          |> Memo.all
+        in
+        List.fold_left
+          (prefix_entries :: subdirs_entries)
+          ~init:Coq_lib_name.Map.empty ~f:Coq_lib_name.Map.union_exn
+      | Error _ ->
+        (* TODO Ignore errors for now *)
+        Memo.return Coq_lib_name.Map.empty)
+    | Error x ->
+      User_error.raise
+        [ Pp.text "System error encountered when finding coqlib:"
+        ; Unix_error.Detailed.pp x
+        ]
+    | _ ->
+      Code_error.raise "subdirectory_map: dir does not exist"
+        [ ("name", Coq_lib_name.to_dyn name); ("dir", Path.to_dyn dir) ]
+
+  let lib_of_user_contrib_name name path : lib =
+    { loc = Loc.none
+    ; boot = Resolve.return None
+    ; id = Id.create ~name ~path
+    ; implicit = false
+    ; use_stdlib = false
+    ; src_root = path
+    ; obj_root = path
+    ; theories =
+        Resolve.return [] (* These may exist but we can't know about them *)
+    ; libraries =
+        Resolve.return [] (* These may exist but we can't know about them *)
+    ; theories_closure =
+        lazy (Resolve.return [])
+        (* These may exist but we can't know about them *)
+    ; package =
+        None
+        (* For now user-contrib entries will not be associated with a package *)
+    }
+
+  let from_coqlib ~coqlib =
+    let open Memo.O in
+    let* stdlib = stdlib_lib ~coqlib in
+    let* subdirs_map =
+      let user_contrib =
+        Path.append_local coqlib (Path.Local.of_string "user-contrib")
+      in
+      subdirectory_map Coq_lib_name.empty user_contrib
+    in
+    let resolve coq_lib_name =
+      let looking_for_stdlib =
+        Ordering.is_eq
+          (Coq_lib_name.compare coq_lib_name (Coq_lib_name.of_string "Coq"))
+      in
+      match looking_for_stdlib with
+      | true -> `Stdlib stdlib
+      | false -> (
+        Coq_lib_name.Map.find subdirs_map coq_lib_name |> function
+        | Some path ->
+          `User_contrib (lib_of_user_contrib_name coq_lib_name path)
+        | None -> `Not_found)
+    in
+    Memo.return { parent = None; resolve; boot = None }
+
+  let installed (context : Context.t) =
+    let open Memo.O in
+    (* First we find coqc so we can query it *)
+    Context.which context "coqc" >>= function
+    | None ->
```

Comment:
There is a tricky case here, when coqc is installed, but coqc is also present in the workspace; in this case we must not use user_contrib.

Reply:
I need to check, does this find coqc in the workspaces? I don't think so right?